### ISSUE TITLE
ValueError: Invalid isoformat string

### ISSUE THREAD

**Issue by test2a** [lullius/pylibby#11]
````
python3 pylibby.py -dl 856947 -f audiobook-mp3 -o /home/user/audiobooks/
Downloading 856947
Loaded archive ./config/archive.json
Traceback (most recent call last):
  File "/home/user/Downloads/pylibby-main/pylibby.py", line 925, in <module>
    L.download_loan(L.get_loan(sys.argv[arg_pos + 1]), args.format, args.output, args.save_info, get_odm=args.odm, embed_metadata=args.embed_metadata, format_string=args.output_format_string, replace_space=args.replace_space, create_opf=args.create_opf)
  File "/home/user/Downloads/pylibby-main/pylibby.py", line 639, in download_loan
    self.download_audiobook_mp3(loan, output_path, save_info=save_info, embed_metadata=embed_metadata, format_string=format_string, replace_space=replace_space, create_opf=create_opf)
  File "/home/user/Downloads/pylibby-main/pylibby.py", line 370, in download_audiobook_mp3
    final_path = os.path.join(output_path, self.get_download_path(audiobook_info["media_info"], replace_space=replace_space))
  File "/home/user/Downloads/pylibby-main/pylibby.py", line 316, in get_download_path
    format_string = format_string.replace("%y", str(datetime.datetime.fromisoformat(media_info['publishDate']).year))
ValueError: Invalid isoformat string: '1993-06-30T04:00:00Z'

````

**Comment by test2a** [comment]
another error while downloading another file....

this looks like some network issue but then again, i tried the library loans and that worked out fine so i don't know what is wrong with this

````
python3 pylibby.py -dl 6328753 -f audiobook-mp3 -o /home/user/audiobooks/
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/connection.py", line 174, in _new_conn
    conn = connection.create_connection(
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/util/connection.py", line 95, in create_connection
    raise err
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/util/connection.py", line 85, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 386, in _make_request
    self._validate_conn(conn)
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 1042, in _validate_conn
    conn.connect()
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/connection.py", line 358, in connect
    self.sock = conn = self._new_conn()
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/connection.py", line 186, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPSConnection object at 0x7fb0ee062710>: Failed to establish a new connection: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/requests/adapters.py", line 489, in send
    resp = conn.urlopen(
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 787, in urlopen
    retries = retries.increment(
  File "/home/user/.local/lib/python3.10/site-packages/urllib3/util/retry.py", line 592, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='sentry-read.svc.overdrive.com', port=443): Max retries exceeded with url: /chip/sync (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fb0ee062710>: Failed to establish a new connection: [Errno 111] Connection refused'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/Downloads/pylibby-main/pylibby.py", line 834, in <module>
    L = Libby(args.id_file, code=args.code, archive_path=args.archive)
  File "/home/user/Downloads/pylibby-main/pylibby.py", line 60, in __init__
    if not self.is_logged_in():
  File "/home/user/Downloads/pylibby-main/pylibby.py", line 86, in is_logged_in
    s = self.get_sync()
  File "/home/user/Downloads/pylibby-main/pylibby.py", line 213, in get_sync
    return self.http_session.get("https://sentry-read.svc.overdrive.com/chip/sync").json()
  File "/home/user/.local/lib/python3.10/site-packages/requests/sessions.py", line 600, in get
    return self.request("GET", url, **kwargs)
  File "/home/user/.local/lib/python3.10/site-packages/requests/sessions.py", line 587, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/user/.local/lib/python3.10/site-packages/requests/sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
  File "/home/user/.local/lib/python3.10/site-packages/requests/adapters.py", line 565, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='sentry-read.svc.overdrive.com', port=443): Max retries exceeded with url: /chip/sync (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7fb0ee062710>: Failed to establish a new connection: [Errno 111] Connection refused'))

````

**Comment by test2a** [comment]
the second one is unrelated to the date error but yeah

**Comment by lullius** [comment]
The minimum Python version is 3.11, you are using 3.10. That is why you get `ValueError: Invalid isoformat string: '1993-06-30T04:00:00Z'`. Please also see #1.  If your system Python is not version 3.11 you can check out [pyenv](https://github.com/pyenv/pyenv) or [asdf](https://github.com/asdf-vm/asdf) to help you manage multiple versions. Pipenv should be able to use pyenv and asdf to install the required Python version for you.

I'm not sure why you get `ConnectionRefusedError: [Errno 111] Connection refused`. Please try again with Python 3.11 and see if that fixes it.

**Comment by test2a** [comment]
@lullius 

````
 python3 --version
Python 3.11.0rc1
user@user-81ut:~/Downloads/pylibby-main$ pipenv install
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/packages/__init__.py", line 27, in <module>
    from . import urllib3
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/packages/urllib3/__init__.py", line 8, in <module>
    from .connectionpool import (
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/packages/urllib3/connectionpool.py", line 35, in <module>
    from .connection import (
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/packages/urllib3/connection.py", line 54, in <module>
    from ._collections import HTTPHeaderDict
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/packages/urllib3/_collections.py", line 2, in <module>
    from collections import Mapping, MutableMapping
ImportError: cannot import name 'Mapping' from 'collections' (/usr/lib/python3.11/collections/__init__.py)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/pipenv", line 33, in <module>
    sys.exit(load_entry_point('pipenv==11.9.0', 'console_scripts', 'pipenv')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 697, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/cli.py", line 347, in install
    from .import core
  File "/usr/lib/python3/dist-packages/pipenv/core.py", line 21, in <module>
    import requests
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/__init__.py", line 62, in <module>
    from .packages.urllib3.exceptions import DependencyWarning
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/packages/__init__.py", line 29, in <module>
    import urllib3
ModuleNotFoundError: No module named 'urllib3'
user@user-81ut:~/Downloads/pylibby-main$ pip3 install urllib3
Defaulting to user installation because normal site-packages is not writeable
Collecting urllib3
  Using cached urllib3-1.26.14-py2.py3-none-any.whl (140 kB)
Installing collected packages: urllib3
Successfully installed urllib3-1.26.14
user@user-81ut:~/Downloads/pylibby-main$ pipenv install
Traceback (most recent call last):
  File "/usr/bin/pipenv", line 33, in <module>
    sys.exit(load_entry_point('pipenv==11.9.0', 'console_scripts', 'pipenv')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 697, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/vendor/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/pipenv/cli.py", line 347, in install
    from .import core
  File "/usr/lib/python3/dist-packages/pipenv/core.py", line 21, in <module>
    import requests
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/__init__.py", line 65, in <module>
    from . import utils
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/utils.py", line 27, in <module>
    from .cookies import RequestsCookieJar, cookiejar_from_dict
  File "/usr/lib/python3/dist-packages/pipenv/vendor/requests/cookies.py", line 172, in <module>
    class RequestsCookieJar(cookielib.CookieJar, collections.MutableMapping):
                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'collections' has no attribute 'MutableMapping'
user@user-81ut:~/Downloads/pylibby-main$ 

````

**Comment by lullius** [comment]
You also need an up to date installation of pipenv. See #10.

**Comment by lullius** [comment]
This should be fixed by #12.
